### DISCUSSION TITLE
Enhancements to utils.connectivity_matrix?

### DISCUSSION THREAD

**Discussion by dPys** [dipy/dipy#1665] [Tractography / Tractometry]
Hi there,

At the moment, Dipy quantifies connectivity based on fiber streamline counts where tracking is performed by seeding from all voxels (or random voxels) of a segmented white-matter mask (or FA-thresholded equivalent). I'd like to extend this functionality to do mask-to-mask connectivity with probabilistic tracking, where seeding is restricted to each roi as a seed mask and every other roi as a termination mask. This is akin to the functionality of probtrackx2 in network mode (i.e. where the tracking only keep paths going through at least one of the other seed masks). What would be the easiest way to implement this with Dipy? see some scratch work below

```
label_mask_seeds = []
for parcel_data in masked_parcel_list:
    label_mask_seeds.append(utils.seeds_from_mask(parcel_data, density=5, affine=dwi_img.get_affine()))

streamlines_list = []
for seed_mask in label_mask_seeds:
    streamline_generator = LocalTracking(dg, act_classifier, seed_mask, affine=dwi_img.affine, step_size=0.5, maxlen=200)
    streamlines_list.append(Streamlines(streamline_generator))

conn_matrices = []
for streamlines in streamlines_list:
    tracks = [sl for sl in streamlines if len(sl) > 1]
    conn_matrix, grouping = utils.connectivity_matrix(tracks, labels_data, affine=labels_affine,
                                                      return_mapping=True, mapping_as_streamlines=True, symmetric=True)
    conn_matrices.append(conn_matrix)

final_mat = np.mean(conn_matrices)
```

Cheers,
Derek

**Comment by arokem** [comment]
Hi Derek - apologies for the slowness here. 

This looks fine to me. Does it work? 😄 

Just to understand what you are trying to do here: why not do one round of whole-brain tracking and then filter post-hoc to find the streamlines that fulfill the conditions (e.g., using [select_by_rois](https://github.com/nipy/dipy/blob/master/dipy/tracking/streamline.py#L344))?

**Comment by dPys** [comment]
Hi @arokem ,

Apologies for my slowness now!

The snippets I had sent actually did not work. But these (sort of) do:

```
def run_tracking(seed_mask, prob_dg, act_classifier, dwi_img, dwi_aligned_atlas):
    from dipy.tracking import utils
    from dipy.tracking.streamline import Streamlines
    from dipy.tracking.local import LocalTracking
    print("%s%s%s" % ('Tracking from: ', seed_mask, '...'))
    atlas_data = nib.load(dwi_aligned_atlas).get_data().astype('int')
    streamline_generator = LocalTracking(prob_dg, act_classifier, seed_mask, affine=dwi_img.affine, step_size=0.5, maxlen=200, return_all=True)
    streamlines = Streamlines(utils.target(Streamlines(streamline_generator), atlas_data,
                                           affine=streamline_generator.affine, include=True))
    return streamlines

class tracking(object):
    def __init__(self, dwi_in, nodif_B0_mask, gm_mask_in_dwi, vent_csf_in_dwi, wm_mask_in_dwi, wm_mask_in_dwi_bin, fbval_in, fbvec_in, dwi_aligned_atlas):
        self.dwi = dwi_in
        self.nodif_B0_mask = nodif_B0_mask
        self.gm_mask_in_dwi = gm_mask_in_dwi
        self.vent_csf_in_dwi = vent_csf_in_dwi
        self.wm_mask_in_dwi = wm_mask_in_dwi
        self.wm_mask_in_dwi_bin = wm_mask_in_dwi_bin
        self.fbval = fbval_in
        self.fbvec = fbvec_in
        self.dwi_aligned_atlas = dwi_aligned_atlas

    def estimate_mod(self):
        from dipy.reconst.csdeconv import ConstrainedSphericalDeconvModel, recursive_response
        from dipy.tracking.local import ActTissueClassifier
        from dipy.tracking import utils, streamline
        from dipy.direction import ProbabilisticDirectionGetter
        from dipy.data import default_sphere
        from dipy.core.gradients import gradient_table
        from dipy.io import read_bvals_bvecs
    
        self.dwi_img = nib.load(self.dwi)
        self.data = self.dwi_img.get_data()

        # Loads mask and ensures it's a true binary mask
        self.mask_img = nib.load(self.nodif_B0_mask)
        self.mask = self.mask_img.get_data() > 0
        self.mask = self.mask > 0

        # loading bvecs/bvals
        self.bvals, self.bvecs = read_bvals_bvecs(self.fbval, self.fbvec)

        # Creating the gradient table
        self.gtab = gradient_table(self.bvals, self.bvecs)
    
        # Load tissue maps and prepare tissue classifier
        # self.gm_mask = nib.load(self.gm_mask_in_dwi)
        # self.vent_csf_mask = nib.load(self.vent_csf_in_dwi)
        # self.wm_mask = nib.load(self.wm_mask_in_dwi)
        # self.background = np.ones(self.gm_mask.shape)
        # self.background[(self.gm_mask.get_data() + self.wm_mask.get_data() + self.vent_csf_mask.get_data()) > 0] = 0
        # self.include_map = self.gm_mask.get_data()
        # self.include_map[self.background > 0] = 1
        # self.exclude_map = self.vent_csf_mask.get_data()
        # self.act_classifier = ActTissueClassifier(self.include_map, self.exclude_map)
        self.vent_csf_mask = nib.load(self.vent_csf_in_dwi).get_data().astype('bool')
        self.wm_mask_dil = nib.load(self.wm_mask_in_dwi_bin).get_data().astype('bool')
        self.act_classifier = ActTissueClassifier(self.include_map, self.exclude_map)

        print('Fitting CSD model...')
        # Instantiate recursive response
        self.response = recursive_response(self.gtab, self.data, mask=self.wm_mask_dil, sh_order=8, peak_thr=0.01,
                                           init_fa=0.08, init_trace=0.0021, iter=8, convergence=0.001, parallel=True)
        # Instantiate CSD
        self.csd_model = ConstrainedSphericalDeconvModel(self.gtab, self.response)
        # Fit CSD
        self.csd_fit = self.csd_model.fit(self.data, mask=self.wm_mask_dil)
        # Grab fod
        self.fod = self.csd_fit.odf(default_sphere)
        # Clip PMF
        self.pmf = self.fod.clip(min=0)
        # Instantiate Probabilistic Estimator
        try:
            # First attempt using the spherical harmonic
            self.prob_dg = ProbabilisticDirectionGetter.from_pmf(self.pmf.shm_coeff, max_angle=30., sphere=default_sphere)
        except:
            # Fallback to entire PMF
            self.prob_dg = ProbabilisticDirectionGetter.from_pmf(self.pmf, max_angle=30., sphere=default_sphere)
        return

    def gen_seeds(self):
        from dipy.tracking import utils
        seed_dens = 5
        self.atlas_img = nib.load(self.dwi_aligned_atlas)
        self.atlas_data = self.atlas_img.get_data().astype('int')
        [self.coords, _, _, self.masked_parcel_list] = get_names_and_coords_of_parcels(self.dwi_aligned_atlas)
        # For each atlas parcel, generate seeds
        self.label_mask_seeds = []
        for parcel_img in self.masked_parcel_list:
            self.label_mask_seeds.append(utils.seeds_from_mask(parcel_img.get_data(), density=seed_dens,
                                                               affine=parcel_img.get_affine()))
        return

    def tractography(self):
        self.streamlines_list = []
        for seed_mask in self.label_mask_seeds:
            self.streamlines_list.append(run_tracking(seed_mask, self.prob_dg, self.act_classifier, self.dwi_img, self.dwi_aligned_atlas))
        return

    def connectometry(self):
        self.tracks_arr = nib.streamlines.ArraySequence([np.concatenate(streamlines) for streamlines in self.streamlines_list])
        self.conn_matrix, self.grouping = utils.connectivity_matrix(self.tracks_arr, self.atlas_data,
                                                                  affine=self.dwi_img.affine, return_mapping=True,
                                                                  mapping_as_streamlines=True, symmetric=True)
        return self.conn_matrix
```

In response to your question, post-hoc filtering of streamlines is certainly a valid approach that can be taken for determinstic tractography, but I'm not sure I see how the results would be any different when using probabilistic tractography? From my understanding, to see any gains from probabilistic tracking, the connectome labels/rois need to be involved in the reconstruction itself, and with repeated sampling so as to make optimal use of the PMF?

Although seed mask/termination mask tracking between each label ROI with every other label ROI may not be the most efficient strategy, my experiences with probtrackx2 have been that the results are far denser, yet sparser (i.e. ~optimal). In the least, I would think that iterating across the atlas labels as seed masks (each time using the entire atlas as a termination mask) would provide a compromise between both extremes?

Curious to get your thoughts and any suggestions to improve my existing code. fyi: I'm working with vogo et al. on this, which is in fact a new prototype for NDMG/PyNets. We are also considering mrtrix3, but would prefer to get this working with Dipy.

@dPys
